test/06-device_log.ts

```diff
@@ -0,0 +1,184 @@
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([createLog({ hello: 'hi there' })])
+			.expect(201));
+
+	it('should reject batches with broken logs', async () => {
+		for (const extra of [
+			{ message: 123 },
+			{ message: null },
+			{ timestamp: new Date().toISOString() },
+		]) {
+			await supertest(app, ctx.device.apiKey)
+				.post(`/device/v2/${ctx.device.uuid}/logs`)
+				.send([createLog(extra)])
+				.expect(400);
+		}
+	});
+
+	// Reading Logs
+
+	it('should allow users to read device logs with a JWT', async () => {
+		const res = await supertest(app, ctx.user)
+			.get(`/device/v2/${ctx.device.uuid}/logs`)
+			.expect(200);
+
+		expect(res.body).to.have.lengthOf(6);
+		let lastTs = 0;
+		(res.body as AnyObject[]).forEach((log, index) => {
+			expect(log).to.have.property('message').that.equals('a log line');
+			expect(log).to.have.property('isStdErr').that.equals(true);
+			expect(log).to.have.property('isSystem').that.equals(true);
+			expect(log).to.have.property('timestamp').that.is.a('number');
+			expect(log).to.have.property('createdAt').that.is.a('number');
+			// Validate they are sorted chronologically
+			expect(log.createdAt).to.be.gte(lastTs);
+			lastTs = log.createdAt;
+			// The 4th is a service log and should have the correct serviceId
+			if (index === 3) {
+				expect(log).to.have.property('serviceId').that.equals(10);
+			} else {
+				expect(log).not.to.have.property('serviceId');
+			}
+		});
+	});
+
+	it('should allow users to read device logs with user-level API keys', async () => {
+		const res = await supertest(app, ctx.user.apiKey)
+			.get(`/device/v2/${ctx.device.uuid}/logs`)
+			.expect(200);
+		expect(res.body).to.have.lengthOf(6);
+	});
+
+	it('should support the `count` option in the custom read endpoint if available', async () => {
+		const res = await supertest(app, ctx.user)
+			.get(`/device/v2/${ctx.device.uuid}/logs?count=4`)
+			.expect(200);
+		expect(res.body).to.have.lengthOf(4);
+		// Test that it's sending the latest 4 and not just any other 4 logs
+		(res.body as AnyObject[]).forEach((log, index) => {
+			if (index === 1) {
+				expect(log).to.have.property('serviceId').that.equals(10);
+			} else {
+				expect(log).not.to.have.property('serviceId');
+			}
+		});
+	});
+
+	it('should support the `count=all` option in the custom read endpoint if available', async () => {
+		const res = await supertest(app, ctx.user)
+			.get(`/device/v2/${ctx.device.uuid}/logs?count=all`)
+			.expect(200);
+		expect(res.body).to.have.lengthOf(6);
+	});
+
+	it('should reject batches with more logs than allowed', () => {
+		const logs = _.times(11, createLog);
+		return supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send(logs)
+			.expect(400);
+	});
+
+	// TODO: Activate this test again, once the billing model can handle retention limits
+	it.skip("should take the retention limit from the device owner's plan", async () => {
```

Comment:
This can be removed altogether, the open source API will never have billing-related stuff.

Reply:
true, removed thanks